jetpack-mu-wpcom: Add code block

sirreal opened the pull request:
## Proposed changes:

* Add an enhanced Code block to jetpack-mu-wpcom.

The block is disabled by default. It requires a filter to be activated. This PR introduces the implementation but does not enable any new functionality.

The block currently looks like this:

| Inserter | Block |
| --- | --- |
| <img width="756" height="740" alt="Screenshot 2025-09-23 at 11 43 00" src="https://github.com/user-attachments/assets/e1287553-5ef9-493d-9b95-74f2a8277ad6" /> | <img width="1666" height="554" alt="Screenshot 2025-09-23 at 11 43 55" src="https://github.com/user-attachments/assets/8e5f879a-3983-46d3-bc67-63463ddb4c1d" /> |

The block name and some warnings will be changed before final release but are still important to keep during testing.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

See pgsdXZ-aB-p2.

It's important to confirm that:
- The block is not enabled by default.
- No errors or warnings are produced when the block is disabled.

* Test these changes on a WordPress.com simple site. With the exception of a few test sites, the block should not be available.
* Test these changes on a WoA (WordPress.com + WP.Cloud) site. The block should not be available.

Before release, the block will undergo a complete Call for Testing.

The block can be tested with these filters:

```php
// Enable the block
add_filter( 'jetpack_mu_wpcom_should_load_code_block', '__return_true' );

// Disable the block
add_filter( 'jetpack_mu_wpcom_should_load_code_block', '__return_false' );
```

sirreal committed:
Add block

sirreal committed:
Add codemirror and lezer dependencies

sirreal committed:
Add grammar loader

sirreal committed:
changelog

sirreal commented:
> How can I test it?

I'll prepare an easy way to test this soon on WordPress.com. There's also the [Jetpack Beta Tester plugin](https://jetpack.com/download-jetpack-beta/) for testing on WordPress.com WP Cloud sites.


> What's left to do?

I just got it building and working, so I need to review things myself and make sure it's finished and all working correctly 🙂

sirreal committed:
Add TypeScript setup

sirreal committed:
Fix misuse of background instead of text color in editor

sirreal committed:
Fix incorrect string replace

sirreal committed:
Remove unused AttributesConfig

sirreal committed:
format dts

sirreal committed:
Allow console.log

sirreal committed:
Type transform arg

sirreal committed:
Fix no-useless-escape lint

sirreal committed:
Assert globalThis type

sirreal committed:
Fix no-nested-ternary lint

sirreal committed:
Remove biome ignore comments

sirreal committed:
Remove loading debug code

sirreal committed:
Eslint ignore import of exposed script module

sirreal committed:
Satisfy tooling disagreements on dot notation

sirreal committed:
Add that description for the linter

sirreal committed:
Fix older React type errors

sirreal committed:
Add the function description

sirreal committed:
Fix shadowing lint

sirreal committed:
Fix exhaustive deps lint

sirreal committed:
Fix more of the lints!

sirreal committed:
Webpack ignore the raw module import

sirreal committed:
da lints.

sirreal committed:
Fix not-equality check

sirreal committed:
Fix different path separator handling

sirreal committed:
Remove duplicate line

sirreal committed:
Fix typo in color properties

sirreal committed:
Use #private over private annotation

sirreal committed:
Handle non-OK worker fetch responses

sirreal commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block.php:
```diff
@@ -0,0 +1,418 @@
+		);
+		foreach ( $color_attributes as $color_attr ) {
+			if ( ! empty( $attributes[ $color_attr ] ) ) {
+				$style_properties[] = "--{$color_attr}: {$attributes[ $color_attr ]};";
```
That escaping is handled by `get_block_wrapper_attributes` later, but good try 👍

sirreal committed:
Remove redundant empty code check

sirreal committed:
Extract class to its own file

sirreal committed:
Function description lints

sirreal committed:
Fix style attribute to include trailing ";"

sirreal committed:
Fix `false` attributes working as enabled

`showLanguageName` could be set to `false` and show.

sirreal committed:
Remove trailing comma

sirreal committed:
Load in footer

sirreal committed:
Ignore nonsense lints

sirreal committed:
Import language data as virtual webpack module

sirreal committed:
Simplify plugin

sirreal committed:
Remove ignore

sirreal committed:
Use global wp_should_load_block_assets_on_demand function

sirreal committed:
Revert "Use global wp_should_load_block_assets_on_demand function"

This reverts commit a4608f3d8c4e8e229cf41c223c3344a678f8e721.

sirreal committed:
More consistent jetpackWebpackConfig destructuring

sirreal committed:
Rework plugin to produce a virtual module

sirreal committed:
Run prettier on plugin

sirreal committed:
Fix eslint ignore comment

sirreal committed:
Remove unused external link

sirreal committed:
Use proper imports in the script

sirreal committed:
Display a loading placeholder when no code is present

sirreal committed:
Always display languagePicker when selected

Use language-specific filename extension in placeholder when available

sirreal committed:
Fix imports, lints, experimental features in color-tools

sirreal committed:
Fix imports and lints in code block transforms

sirreal committed:
Add missing @wordpress/editor dep

sirreal committed:
Fix worker lints

sirreal committed:
Lints language-utils

sirreal committed:
Disable worker Function type lints

sirreal committed:
Improve all the script loading logic

Use asset files
Use script module data for worker URL

sirreal committed:
Fix block.json textdomain lint

sirreal committed:
Fix CSS lints

sirreal committed:
Fix more lints

sirreal committed:
Fix up some worker-admin types

sirreal committed:
Lint lint lint

sirreal committed:
Fix editor css lints

sirreal committed:
Migrate everything to webpack and asset files with versions

sirreal committed:
Check `function_exists` for wp_should_load_block_assets_on_demand function from WP 6.8+

sirreal committed:
Ignore types not provided by Core

sirreal committed:
Rely on internal types

sirreal committed:
Fix types in color-tools

sirreal committed:
Add phan suppress

sirreal committed:
Fix allowed filename lint

sirreal committed:
Try to suppress that phan error

sirreal committed:
I hope this incantantion is the right one

sirreal committed:
nope

sirreal committed:
It works much better on the right line

sirreal committed:
Fix typo in variable name

sirreal committed:
Remove redundant optional chain

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx:
```diff
@@ -0,0 +1,482 @@
+ *
+ * /Color schemes
+ *
+ */
```
what’s going on with these comments? Why do they have all the extra newlines and why does this one start with a solidus?

sirreal replied:
They were organizational to delimit a section. They're left over from when there was a lot more in this file. I'll remove them.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx:
```diff
@@ -0,0 +1,482 @@
+
+	return (
+		<Chrome isLoading { ...props }>
+			<pre className="cm-content">{ code }</pre>
```
no `<code>` in the wrapper?

sirreal replied:
Good spot. This should likely match the save function in general, although I think this is going to change significantly to match the core/code block.

I'll update.

dmsnell replied:
were there missing updates? I noticed here and in the next comment about extracting a type that you said you updated or will update, but the code is the same. just curious if there was a missed `git add`

sirreal replied:
There was an oversight in the change, this line needed to be deleted. It was causing the Loading code (`Loading…` on a new block) to be duplicated. That's been handled now.

The type change from the other comment was applied. The line where it was commented did not change so it may not be clear in the GitHub UI:

https://github.com/Automattic/jetpack/blob/6e64336150efbab7161a30fd3d1bfb9896493556/projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx#L421-L445

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx:
```diff
@@ -0,0 +1,482 @@
+	[ key in `--${ keyof Pick<
+		Attributes,
+		Extract< keyof Attributes, `color${ Capitalize< string > }` >
+	> }` ]-?: string | undefined;
```
a type this gnarly would benefit by having its own name and documentation.

sirreal replied:
I extracted, improved, and documented this type.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx:
```diff
@@ -0,0 +1,482 @@
+		.replaceAll( '<', '&lt;' )
+		.replaceAll( '>', '&gt;' )
+		.replaceAll( '[', '&#91;' );
+}
```
while not required, I think we tend to escape single and double quotes as well. worth doing here?

sirreal replied:
[You are correct, `esc_html` encodes quotes.](https://github.com/WordPress/wordpress-develop/blob/98af3c5b4a60e34080e5d2c1d4030fd9c3cfc45b/src/wp-includes/formatting.php#L4685-L4687) I guess I'll escape them here.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
```diff
@@ -0,0 +1,692 @@
+				const [ , langCandidate ] = CODE_FENCE_REGEXP.exec( content )!;
+
+				let language;
+				switch ( langCandidate?.toLowerCase() ) {
```
this mapping of language aliases to language name warrants its own function. it seems like we’re already using something like it elsewhere, aren’t we?

sirreal replied:
These switches aren't identical. This list largely seeks to match GitHub's code fence behavior. Folks working with snippets have expectations about what **```js** will do.

dmsnell replied:
it makes sense that they aren’t identical, but these are also transforms with very little cost to getting it wrong, right? like maybe on Github if you type `rmb3` it won’t have any idea what to do with that and uses plaintext. here it accidentally selects `Ruby Multi-Byte 3`. oops. you can change to plaintext…

but I find it hard to imagine someone typing `js` in either place and expecting anything other than JavaScript. that’s what I’m suggesting about this mapping of aliases to supported language. we can be broad and accept more than Github or other code blocks support — is that a bad thing?

sirreal replied:
No, not a bad thing and a good suggestion. I'll see about unifying these into a single function.

sirreal replied:
I unified these all into a single function. It seems much better.

There is one special case I added because [SyntaxHighlighter block](https://wordpress.org/plugins/syntaxhighlighter/) uses ['xml' for several languages, including HTML and XML](https://github.com/Automattic/syntaxhighlighter/blob/61fd26c64d019641db143e90e05b0eb9a4a57d62/syntaxhighlighter.php#L240-L243). I opted to transform to HTML in that case.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
case in point: seems like this could/should be merged into the list from above.

sirreal replied:
This switch handles the language attribute for Code Block Pro. There's overlap, but the list and language format are not identical.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
first of all, why are we bailing on files larger than 10 KB?
second of all, why are we doing it _here_ and not in `getLanguage()`?

sirreal replied:
It's an arbitrary number to limit and could be removed or tweaked.

My thinking was that a small-ish file is more likely to be desirable as a code block. Large files are less likely to be code snippets and more likely to be file uploads. This was just an assumption of mine.

**I'm not even sure this transform should exist.** When I've used it, it's a joy, but I expecting it. As this block moves out of a standalone plugin and is included for a larger group of folks, it's possible that this violates their expectations of drag and drop behavior.

dmsnell replied:
the image block already captures drag and drop. I think it could be a great feature.

perhaps we could examine the `file.type` and see if it’s plaintext, in addition to or instead of looking at the file extension, to drop it in. I wonder if it’s possible to show a preview on hover that would clue someone in to the fact that if they let go, the text goes into the code block.

sirreal replied:
That's a good idea and is potentially a better solution than the extension matching below for some file types that are very likely to be code. I like it!

sirreal replied:
Do you have thoughts on a MIME type handling implementation? I don't really want to implement a MIME parser, and some limited testing on my has been interesting.

- Many files don't have any `type` information (it's empty)
- Some files report expected values, like `application/json` or `text/css`.
- Some files report reasonable things that don't seem to be on the IANA list of MIME types, like `text/php`.
- Some files have IANA defined types `application/yaml` but report `application/x-yaml`.

I don't really want to implement a MIME parser here, I'm wondering if you have clear ideas. Maybe we should just handle some of the most common, official IANA MIME types and let others fall through to be handled by extension.

- `text/xml` and `application/xml`
- `application/json`
- `text/css`
- etc.

I'm consulting this resource: https://www.iana.org/assignments/media-types/media-types.xhtml

dmsnell replied:
if it starts with `text/` that should be fine to use. there are surely a few `application/` types we can use, and if they are defined in the IANA list that would be great.

hadn’t considered it before but this could also be a huge tentative clue for determining the language apart from the filename.

here is a list of types I found which end in `+`:

```
+cbor
+cbor-seq
+cose
+csv
+cwt
+der
+fastinfoset
+gzip
+json
+json-seq
+jwt
+sqlite
+tlv
+wbxml
+xml
+yaml
+zip
```

to me it seems reasonable to have some basic rules:
 - if it reports `text/` then accept it
 - if it ends in `+csv`/`+json`/`+json-seq`/`+jwt`/`+xml`/`+yaml` accept it.

this is all a best-effort system with low chance of messing up. seems like this would be _reasonable_ without having a spec-compliant parser for the MIME types and without understanding all of them.

sirreal replied:
I pushed a basic implementation to recognized some well known IANA MIME types.

I thought about using the `+` part to pick a language, like `+xml` uses `XML` language, however I wonder if that's always applicable. To pick an example, `vnd.zzazz.deck+xml` is a `.zaz` file used by the Zzazz application. Is it likely folks want to display that in a Code block?

The simple handling of a select known cases may be more appopriate.

dmsnell replied:
> Is it likely folks want to display that in a Code block?

If they drag it into the block I would expect so. I’m not going to stick on this, it’s not a big point.

What I think is clear is that it’s helpful to show that the file is being dragged onto the block, because that lets the user decide. If they _do_ drop it onto the code block, then I think there’s very little reason not to understand that `+xml` means it’s XML.

sirreal replied:
This transform handler is when a file is dragged onto the Block Editor canvas. Not actually inside a Code block.

Testing on my system, when certain files are dropped onto an existing code block, the test from the file is inserted into the code block at the cursor location. [I believe that's core functionality of `@codemirror/view`](https://github.com/codemirror/view/blob/5559e0052d5b1e6601f01bdce558b53be4caa34f/src/input.ts#L680-L696), nothing specific has been implemented in the block to handle that.

sirreal replied:
Do you have thoughts about the 10KB size limit and whether it should remain or be adjusted?

dmsnell replied:
as for _any_ size limit I don’t have strong feelings, but maybe _some_ limit is reasonable. in my experience though 10 KB feels too small. for instance, I have posted and seen people post log files, even inside a `<details>` element. with only 100 lines that’s less than 100 bytes per line.

if I were to guess, I would think that rendering performance could be more of guide than semantics. at what point does CodeMirror start choking? 100 KB? probably not. 1 MB? maybe at this point it’s unreasonable to use.

if drag and drop fails then someone can always copy and paste.

sirreal replied:
I don't mind removing the limit, it was an quick decision of mine. Not a lot of thought went into it.

sirreal replied:
I increased the limit to 1MB.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
MDN claims that `name` has the path removed for security purposes, so it seems like this code should never find a legitimate path, meaning that the only time we expect it to run, it will corrupt a filename.

> For security reasons, the path is excluded from this property. [MDN](https://developer.mozilla.org/en-US/docs/Web/API/File/name)

Also, isn’t it a risk on its own to blur the separators inside a group like this? Regardless of which separator is in use, it will split on the wrong ones.

It seems like the Web Platform hands us the normative file name here so I am having trouble figuring out why we’re attempting to split it and look for it.

sirreal replied:
You're correct, I'll remove the splitting.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
in this case it seems like Copilot detected a proper fix. why force `null.toLowerCase()` when we might expect files to lack a file extension?

also, do we expect any reasonable length or limit on the extension? I have seem some filenames which have a dot early in their name and then it trips up detection like this. `my.file.export.2025-11-23`

sirreal replied:
I don't believe this is nullable, which is why I used the non-null assertion. There shouldn't be a way to split a string into an empty array. And a non-empty array will always have a string as the last (-1th) element.

This is a place where the types are correct but don't faithfully represent the behavior. I'll justify the non-null assertion in a comment.

- `split` returns `string[]`. Any input will return at least `[ string, ...string[] ]`.
- `at` returns `string | undefined`. But [ "string" ].at(-1) will never return undefined.

---

> do we expect any reasonable length or limit on the extension? I have seem some filenames which have a dot early in their name and then it trips up detection like this. `my.file.export.2025-11-23`

This is a reasonable best effort. If it looks like something we can represent as a code block, we do. Otherwise it will fall through to be handled by another handler, likely the file block. I think it's reasonable to expect a file to have a recognizable extension (like `.json`) in order to trigger this behavior.

dmsnell replied:
both points make sense. thanks for pointing out the split thing. that did not jump out at me.

dmsnell reviewed (commented):
Unfortunately I had to stop reviewing this because Github’s UI has become unusuably slow. I can’t scroll without it jumping far down the page and it’s taking me sometimes over 12 seconds to open a line-comment box or post one.

gziolo commented:
> The block is not enabled by default.
> No errors or warnings are produced when the block is disabled.
> Test these changes on a WordPress.com simple site. With the exception of a few test sites, the block should not be available.

I confirm it behaves as expected.

sirreal committed:
Remove section delimiter comments

sirreal committed:
Fix code block wrapper in loading

sirreal committed:
Escape quote characters

sirreal committed:
Remove attempted file name path splitting

sirreal committed:
Extract complex color custom properties type

Extract the type, complete it, and add explanatory comments.

sirreal committed:
Add comment justifying non-null assertion

sirreal committed:
Add basic MIME file.type switch on transform

sirreal commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/block-definition.tsx:
```diff
@@ -0,0 +1,517 @@
+		.replaceAll( '>', '&gt;' )
+		.replaceAll( '[', '&#91;' )
+		.replaceAll( "'", '&#39;' )
+		.replaceAll( '"', '&#22;' );
```
`22` is in hex, this should be the decimal `34`:
```suggestion
		.replaceAll( '"', '&#34;' );
```

sirreal committed:
Fix quote character decimal hex entity.

sirreal committed:
Improve safety of HTML replacement and escaping

sirreal committed:
Rename getLanguageFromFile function

sirreal committed:
Unify the language recognition switches into a single function

sirreal commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block-html-replacer.php:
```diff
@@ -0,0 +1,123 @@
+				$code = base64_decode( $chunk[0], true );
+				if ( false === $code ) {
+					return null;
+				}
+
+				$class_name = $chunk[1] ?? null;
+
+				/*
+				 * Do not rely on `esc_html`, it would mishandle character references
+				 * that appear to be encoded already. HTML like `&amp;` would be
+				 * ignored, resulting in `&` rendering in the browser instead of the
+				 * desired HTML `&amp;` which must be encoded as `&amp;amp;`.
+				 *
+				 * Below, the `double_encode` argument is set to `true` to ensure
+				 * prevent this issue and ensure correct encoding.
+				 */
+				$html_encoded_code = _wp_specialchars(
+					wp_check_invalid_utf8( $code ),
+					ENT_NOQUOTES,
+					false,
+					true
+				);
```
@dmsnell I'd love your thoughts here. [This is basically what `esc_html` does](https://github.com/WordPress/wordpress-develop/blob/70d00508e950f0e63de82386d453bca5ad899474/src/wp-includes/formatting.php#L4685-L4700) (without the filter applied), but we can't rely on `esc_html` because its double encdode prevention is not optional.

The UTF8 check is interesting and some [recent work of yours has](https://core.trac.wordpress.org/changeset/60793) has made it clear how peculiar the `wp_check_invalid_utf8` function is.

What do you think is best here, should `wp_scrub_utf8()` be used if available, with `wp_check_invalid_utf8()` as a fallback?

dmsnell replied:
tough call @sirreal. part of me thinks we should just rely on `esc_html()`, part of me thinks we should all of this and do manual replacement. `esc_html()` runs through a filter, which itself is suspect, but maybe that’s important to someone.

I would hate to see us add more examples of calling `wp_check_invalid_utf8()` since that function is inherently broken.

sirreal replied:
Reflecting on this, WordPress sanitization and escaping functions are not helping. I'm going to use `htmlspecialchars` directly unless you can think of a reason not to.

sirreal replied:
I went with this for the HTML escaping:

```php
$html_encoded_code = htmlspecialchars( $code, ENT_NOQUOTES | ENT_SUBSTITUTE );
```

That will encode `<`, `&` and `>` and will escape things that already appear to be character references, the desired behavior.

sirreal committed:
Fix typo in color tools eslint comment

sirreal committed:
Remove redundant Number() cast

sirreal committed:
Improve externalLanguage… function doc

sirreal committed:
Check for navigator.clipboard availability

sirreal committed:
Improve innerBlocks type

sirreal committed:
Fix some fx bindings in webpack plugin

sirreal committed:
Use better dev number 😉

sirreal committed:
Remove langNames from bundle, simplify

sirreal committed:
Support both selects

sirreal committed:
Remove duplicate loading code

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-definition/transforms.ts:
```diff
@@ -0,0 +1,531 @@
+						code,
+						triggerCodeUpdate: true,
+					} );
+				} );
```
nine seconds to open this line-comment box 🤦‍♂️, Github…

do you think we should try to avoid wiping out any existing content? I know it’s probably a rare situation, but in the time between loading the file and finishing the load of the file we could have different code in there.

maybe at least we could leave a note for ourselves that this can happen? I suppose that by the point the file is uploaded into the browser we’re never talking about many more than a few milliseconds, but plausibly it could come from a slow external drive or something.

options:
	- capture the pre-upload code and check if it’s the same
	- insert the uploaded file into the cursor position
	- leave a note indicating the content was changed?

sirreal replied:
In my experience this is always immediate, but it's not a bad idea to to handle async behavior better.

Shall I show the loading view while `triggerCodeUpdate` is set?

sirreal replied:
> Shall I show the loading view while `triggerCodeUpdate` is set?

That won't work well. I'll work on a change to address the async nature of this without showing a usable editor.

dmsnell replied:
It does not seem like we would expect delays on the order of showing a loading view. Typing and pasting can be surprisingly fast at times, but I wouldn’t want to show a loading indicator if we expect something to happen within a couple hundred milliseconds.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block-html-replacer.php:
```diff
@@ -0,0 +1,123 @@
+					return null;
+				}
+
+				// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.obfuscation_base64_decode
```
what if we added new rules in our PHPCS config to outlaw calling all functions which in various ways could be part of systems that are used maliciously or accidentally to do harm?

let’s ban PHP!

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block-html-replacer.php:
```diff
@@ -0,0 +1,123 @@
+				}
+
+				// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.obfuscation_base64_decode
+				$code = base64_decode( $chunk[0], true );
```
it might be worth ensuring that we don’t have URL-safe variant issues here, whereas we are encoding and decoding across different systems.

the things to check are whether the editor can produce `-` and `_` or end without the padding characters.

俿 (U+4FFF) as UTF-8 encodes in base64 as: `5L+/`
 - in PHP ✅
 - in `btoa( [ ...( new TextEncoder() ).encode( '俿' ) ].map( i => String.fromCharCode( i ) ).join( '' ) )` ✅
 - in `(new TextEncoder() ).encode( '俿' ).toBase64()` ✅

so as long as we’re not doing something different in the JS code, this should be fine. if we try decoding the alternate version, `5L-_` then PHP rejects it.

sirreal replied:
This is the base64 encoding on the client:

https://github.com/Automattic/jetpack/blob/928fa2c1f3083076e922f319519e59e40679dba2/projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/common/base-64.ts#L14-L24

I think it aligns with OK usage in your testing.

dmsnell commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block.php:
```diff
@@ -0,0 +1,330 @@
+		$replacement_code_data = $attributes['tokenizedLines'];
+
+		$replacer                             = Code_Block_HTML_Replacer::create_fragment( $content );
+		list($code_string, $replaced_content) = $replacer->get_updated_html_with_replaced_content( $replacement_code_data );
```
I think that if we make this all one static method on `Code_Block_HTML_Replacer` then we can create an HTML Processor within that method and access its protected properties and methods from the `$processor` it creates. this is a neat static method of subclass trick. then it wouldn’t require the more awkward instantiation and method call.

```php
class Replacer extends WP_HTML_Processor {
  public static function replace_code_block( $html ) {
    $p = WP_HTML_Processor::create_fragment( $html );

    $p->next_tag( 'code' );
    $p->set_bookmark( 'here' );
    var_dump( "CODE element at {$p->bookmarks['_here']->start}." );
  }
}

Replacer::replace_code_block( '<div><pre><code>This is code</code></pre></div>' );
```

[WP PHP Playground](https://playground.wordpress.net/php-playground.html#eyJjb2RlIjoiPD9waHBcblxucmVxdWlyZSAnL3dvcmRwcmVzcy93cC1sb2FkLnBocCc7XG5cbmNsYXNzIFJlcGxhY2VyIGV4dGVuZHMgV1BfSFRNTF9Qcm9jZXNzb3Ige1xuICBwdWJsaWMgc3RhdGljIGZ1bmN0aW9uIHJlcGxhY2VfY29kZV9ibG9jayggJGh0bWwgKSB7XG4gICAgJHAgPSBXUF9IVE1MX1Byb2Nlc3Nvcjo6Y3JlYXRlX2ZyYWdtZW50KCAkaHRtbCApO1xuXG4gICAgJHAtPm5leHRfdGFnKCAnY29kZScgKTtcbiAgICAkcC0+c2V0X2Jvb2ttYXJrKCAnaGVyZScgKTtcbiAgICB2YXJfZHVtcCggXCJDT0RFIGVsZW1lbnQgYXQgeyRwLT5ib29rbWFya3NbJ19oZXJlJ10tPnN0YXJ0fS5cIiApO1xuICB9XG59XG5cblJlcGxhY2VyOjpyZXBsYWNlX2NvZGVfYmxvY2soICc8ZGl2PjxwcmU+PGNvZGU+VGhpcyBpcyBjb2RlPC9jb2RlPjwvcHJlPjwvZGl2PicgKTsiLCJwaHAiOiI4LjQiLCJ3cCI6IjYuOCJ9)

sirreal replied:
Applied in aabbb364946ea7a439d36b207564b18100d72ffb.

sirreal committed:
Update entity encoding to use htmlspecialchars

sirreal committed:
Use static method to simplify call

sirreal committed:
Ensure tokenizedLines attribute is an array

sirreal committed:
Increase dropped file size allowance to 1MB

gziolo commented on projects/packages/jetpack-mu-wpcom/package.json:
```diff
@@ -94,6 +99,7 @@
 		"@babel/preset-react": "7.27.1",
 		"@babel/runtime": "7.27.6",
 		"@playwright/test": "1.55.1",
+		"@lezer/generator": "1.8.0",
```
Linter most likely wants to see `@lezer/generator` one line above.

gziolo commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/class-code-block.php:
```diff
@@ -0,0 +1,418 @@
+	 * The dummy module is only used to signal that some additional modules
+	 * should be included in the importmap.
+	 *
+	 * @TODO: Be safer. Check the return (bool: was removed) and behave accordingly.
```
Do you plan to address the TODO included? What would get handled differently?

sirreal replied:
Yes, I'll address that in a follow-up PR.

It's possible that the actions aren't removed or have been modified by other plugins. This should just bail and allow the empty module to stay enqueued.

sirreal replied:
Follow-up is in https://github.com/Automattic/jetpack/pull/45357.

gziolo commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-blocks/code/block-front/block-front.ts:
```diff
@@ -0,0 +1,33 @@
+	}
+}
+
+if ( navigator.clipboard && ! document.body.classList.contains( 'wp-admin' ) ) {
```
What does it make incompatible with the Interactivity API?

sirreal replied:
The first implementation used the Interactivity API. Then I realized two things:

- I want this to work correctly on sites that don't render everything on the initial page load (o2/p2 sites, sites with infinite scroll, etc.). The Interactivity API doesn't support that, it performs hydration once on page load.
- I was using the Interactivity API to attach a single click handler. It was actually more complicated and using a library seemed quite excessive.

I decided to implement this as a JavaScript module with no dependencies.

gziolo replied:
Good points. Thank you for explaining your reasoning 👍

gziolo reviewed (dismissed):
Let's proceed with the initial call for testing to check the user experience. In this view it's really difficult to verify which code changes are specific to integration with Jetpack MU wpcom.

sirreal committed:
Expand on htmlspecialchars usage

Use explicit UTF-8 encoding and true double-encode argument.

sirreal committed:
Sort package.json devDeps

sirreal committed:
Fix react warning spread passing key prop

select and custom select use different options object and spread pass props through. prevent a "key" prop warning

sirreal committed:
Clean up and simplify WP types for block edit function
